caffeine@patapon.info/mprisMediaPlayer2.js

```diff
@@ -0,0 +1,342 @@
+// @ts-check
+// @ts-expect-error
+import Gio from 'gi://Gio';
+// @ts-expect-error
+import GObject from 'gi://GObject';
+
+// eslint-disable-next-line
+'use strict';
+
+/**
+ * Represents the DBus proxy class instance.
+ * @typedef {{
+ *  ListNamesRemote(callbackFn: (data: [string[]]) => never): never;
+ *  ListNamesSync(): [string[]];
+ *  ListNamesAsync(): Promise<[string[]]>;
+ *  connectSignal(signal: string, callbackFn: (proxy, sender, [name, oldOwner, newOwner]) => void): any
+ *  disconnectSignal(handlerId: any): void
+ *  }} DBusProxy
+ */
+/**
+ * Represents the DBus proxy class.
+ * @typedef {{
+ *  new(
+ *      bus: string,
+ *      name: string,
+ *      objectPath: string,
+ *      proxy: (proxy: DBusProxy) => void): DBusProxy;
+ * }} DBusProxyClass
+ */
+const DBusInterface = `<node>
+    <interface name="org.freedesktop.DBus">
+        <method name="ListNames">
+            <arg type="as" direction="out" />
+        </method>
+        <signal name="NameAcquired">
+            <arg type="s"/>
+        </signal>
+    </interface>
+</node>`;
+
+/**
+ * Represents the DBus Mpris Player proxy class instance.
+ * @typedef {{
+ *  PlaybackStatus: string;
+ *  connect(signal: string, callbackFn: (player: DBusMprisPlayerProxy) => void): any
+ *  disconnect(handlerId: number): void
+ * }} DBusMprisPlayerProxy
+ */
+/**
+ * Represents the DBus Mpris Player proxy class.
+ * @typedef {{
+ *  new(
+ *      bus: string,
+ *      name: string,
+ *      objectPath: string,
+ *      player: (player: DBusMprisPlayerProxy) => void): DBusMprisPlayerProxy;
+ * }} DBusMprisPlayerProxyClass
+ */
+const DBusMprisPlayerInterface = `<node>
+  <interface name="org.mpris.MediaPlayer2.Player">
+    <property name="PlaybackStatus" type="s" access="read"/>
+  </interface>
+</node>`;
+
+class PrivateContructorParams {
+    dont = 'doit';
+}
```

Comment:
What's this for?

Reply:
Private constructor logic.

https://github.com/PVermeer/gnome-shell-extension-caffeine/blob/b1ec5df0c509f75d9a56a44ed484260a195fc4c1/caffeine%40patapon.info/mprisMediaPlayer2.js#L298-L303

I implemented MPRIS as a singleton service/plugin to this extension. To avoid leaks I made it a singleton class that should be manually constructed. It's just an object thats only available in this module so the class can only be constructed via static methods.

Reply:
Is this necessary since the class won't be accessed outside of Caffeine?

Personally, I don't think the class needs to be as defensive, since it's documented that `Get()` and `Destroy()` should be used:

https://github.com/PVermeer/gnome-shell-extension-caffeine/blob/b1ec5df0c509f75d9a56a44ed484260a195fc4c1/caffeine%40patapon.info/mprisMediaPlayer2.js#L328

Reply:
It's just a safety measure. In other languages one can simple define the constructor private. It doesn't hurt or  effect anything besides accidental class instantiating in future work.

You can change it if you like ofc

Reply:
I removed this logic